src/lib/api-keys.ts

```diff
@@ -1,6 +1,74 @@
+import type { Request } from 'express';
 import * as _ from 'lodash';
+import { errors } from '@balena/pinejs';
+import {
+	createApiKey,
+	ApiKeyOptions,
+	PartialCreateKey,
+} from '../platform/api-keys';
+
+const { BadRequestError } = errors;
+
+const supportedActorTypes = ['application', 'user', 'device'] as const;
+
+export type ApiKeyParameters = {
+	actorType: typeof supportedActorTypes[number];
+	actorTypeId: number;
+	roles: string[];
+} & Pick<ApiKeyOptions, 'name' | 'description' | 'apiKey'>;
+
+export const createGenericApiKey = async (
+	req: Request,
+	{
+		actorType,
+		actorTypeId,
+		roles,
+		name,
+		description,
+		apiKey,
+	}: ApiKeyParameters,
+) => {
+	if (!supportedActorTypes.includes(actorType)) {
+		throw new BadRequestError('Unsupported actor type');
+	}
+
+	if (!Number.isFinite(actorTypeId)) {
+		throw new BadRequestError('Actor must be a number');
+	}
+
+	if (!Array.isArray(roles) || roles.some((r) => typeof r !== 'string' || !r)) {
+		throw new BadRequestError('Roles should be an array of non-empty strings');
+	}
+
+	if (roles.length !== 1) {
+		throw new BadRequestError('API Keys currently only support a single role');
+	}
+
+	if (!name) {
+		const namedRole = roles.find((r) => r.startsWith('named-'));
+
+		if (namedRole != null) {
+			throw new BadRequestError(
+				`API keys with the ${namedRole} role require a name`,
```

Comment:
I like to quote the variable bit in case it includes spaces or something, ie
```suggestion
				`API keys with the '${namedRole}' role require a name`,
```